src/cli/stats-docs.ts

```diff
@@ -0,0 +1,141 @@
+    process.stderr.write(colors[color](`${metric}: ${total} \n`));
+  }
+}
+
+function printStatsJson(statsCount: IStatsCount) {
+  const json: any = {};
+  Object.keys(statsCount).forEach((key) => {
+    json[key] = {
+      metric: statsCount[key as StatsName].metric,
+      total: statsCount[key as StatsName].total,
+    }
+  })
+  process.stdout.write(JSON.stringify(json, null, 2));
+}
+
+function printStats(statsCount: IStatsCount, entrypoint: string, format: string) {
+  process.stderr.write(`Document: ${colors.magenta(entrypoint)} stats:\n\n`);
+  switch (format) {
+    case 'stylish': printStatsStylish(statsCount); break;
+    case 'json': printStatsJson(statsCount); break;
+  }
+}
+
+function printExecutionTime(startedAt: number, entrypoint: string) {
+  const elapsed = getExecutionTime(startedAt);
+  process.stderr.write(colors.gray(`\n${entrypoint}: stats processed in ${elapsed}\n\n`));
+}
+
+export async function handleStats (argv: {
+  config?: string;
+  entrypoint?: string;
+  format: string;
+}) {
+  const config: LintConfig | Config = await loadConfig(argv.config);
+  const entrypoints = await getFallbackEntryPointsOrExit(argv.entrypoint ? [argv.entrypoint] : [], config);
+  const entrypoint = entrypoints[0];
+  const externalRefResolver = new BaseResolver(config.resolve);
+  const document = (await externalRefResolver.resolveDocument(null, entrypoint)) as Document;
+  const lintConfig: LintConfig = config.lint;
+  const oasVersion = detectOpenAPI(document.parsed);
+  const oasMajorVersion = openAPIMajor(oasVersion);
+  const types = normalizeTypes(
+    lintConfig.extendTypes(
+      oasMajorVersion === OasMajorVersion.Version3 ? Oas3Types : Oas2Types,
+      oasVersion,
+    ),
+  );
+
+  const startedAt = performance.now();
+  const ctx: WalkContext = {
+    problems: [],
+    oasVersion: oasVersion,
+  }
+
+  const resolvedRefMap = await resolveDocument({
+    rootDocument: document,
+    rootType: types.DefinitionRoot,
+    externalRefResolver,
+  });
+
+  const statVisitor: Oas3Visitor | Oas2Visitor = {
+    ExternalDocs: { leave() { statsCount.externalDocs.total++; }},
+    ref: { enter(ref: OasRef) { statsCount.refs.items!.add(ref['$ref']); }},
+    Tag: { enter() { statsCount.tags.total++; }},
+    Link: { leave(link: any) { statsCount.links.items!.add(link.operationId); }},
+    PathMap: {
+      leave() {
+        statsCount.parameters.total = statsCount.parameters.items!.size;
+        statsCount.refs.total = statsCount.refs.items!.size;
+        statsCount.links.total = statsCount.links.items!.size;
+      },
+      PathItem: {
+        leave() { statsCount.pathItems.total++; },
+        Operation: { leave() { statsCount.operations.total++; }},
+        Parameter: { leave(parameter: Oas2Parameter | Oas3Parameter) {
+          statsCount.parameters.items!.add(parameter.name)
+        }}
+      }
+    },
+    Components: {
+      Schema: { leave() { statsCount.schemas.total++; }}
```

Comment:
It won't catch OAS2 schemas: https://github.com/Redocly/openapi-cli/blob/master/src/types/oas2.ts#L16

Also, I would just use `NamedSchema` for OAS3 too. It should count all the top-level named schemas but can be defined on the top-level (no need to use nested visitor)